packages/focusvisible/src/useFocusVisible.js

```diff
@@ -0,0 +1,255 @@
+    e => {
+      if (!isValidFocusTarget(e.target)) {
+        return;
+      }
+
+      if (isFocused(e.target)) {
+        hadFocusVisibleRecently.current = true;
+
+        clearTimeout(hadFocusVisibleRecentlyTimeout.current);
+        hadFocusVisibleRecentlyTimeout.current = setTimeout(() => {
+          hadFocusVisibleRecently.current = false;
+          clearTimeout(hadFocusVisibleRecentlyTimeout.current);
+        }, 100);
+
+        removeFocusVisibleClass(e.target);
+      }
+    },
+    [isFocused, isValidFocusTarget, removeFocusVisibleClass]
+  );
+
+  const onInitialPointerMove = useCallback(
+    e => {
+      if (e.target.nodeName && e.target.nodeName.toLowerCase() === 'html') {
+        return;
+      }
+
+      hadKeyboardEvent.current = false;
+      // eslint-disable-next-line no-use-before-define
+      removeInitialPointerMoveListeners();
+    },
+    // eslint-disable-next-line no-use-before-define
+    [removeInitialPointerMoveListeners]
+  );
+
+  const addInitialPointerMoveListeners = useCallback(() => {
+    relativeDocument.addEventListener('mousemove', onInitialPointerMove);
+    relativeDocument.addEventListener('mousedown', onInitialPointerMove);
+    relativeDocument.addEventListener('mouseup', onInitialPointerMove);
+    relativeDocument.addEventListener('pointermove', onInitialPointerMove);
+    relativeDocument.addEventListener('pointerdown', onInitialPointerMove);
+    relativeDocument.addEventListener('pointerup', onInitialPointerMove);
+    relativeDocument.addEventListener('touchmove', onInitialPointerMove);
+    relativeDocument.addEventListener('touchstart', onInitialPointerMove);
+    relativeDocument.addEventListener('touchend', onInitialPointerMove);
+  }, [onInitialPointerMove, relativeDocument]);
+
+  const removeInitialPointerMoveListeners = useCallback(() => {
+    relativeDocument.removeEventListener('mousemove', onInitialPointerMove);
+    relativeDocument.removeEventListener('mousedown', onInitialPointerMove);
+    relativeDocument.removeEventListener('mouseup', onInitialPointerMove);
+    relativeDocument.removeEventListener('pointermove', onInitialPointerMove);
+    relativeDocument.removeEventListener('pointerdown', onInitialPointerMove);
+    relativeDocument.removeEventListener('pointerup', onInitialPointerMove);
+    relativeDocument.removeEventListener('touchmove', onInitialPointerMove);
+    relativeDocument.removeEventListener('touchstart', onInitialPointerMove);
+    relativeDocument.removeEventListener('touchend', onInitialPointerMove);
+  }, [onInitialPointerMove, relativeDocument]);
+
+  /* Unable to mock visibilityState in JSDom environment */
+  /* istanbul ignore next */
+  const onVisibilityChange = useCallback(() => {
+    if (relativeDocument.visibilityState === 'hidden') {
+      if (hadFocusVisibleRecently.current) {
+        hadKeyboardEvent.current = true;
+      }
+    }
+  }, [relativeDocument]);
+
+  useEffect(() => {
+    const currentScope = scope && scope.current;
```

Comment:
`useRef` will always guarantee the `{current: [value]}` shape so probably a redundant check here.

```
const currentScope = scope.current;
```

The above would be fine.

Reply:
I think this was added by the eslint auto-fixer. Updated.

I also added in `relativeDocument` and `scope` checks to better support server-rendering environments.